src/exchange/checkpoint.py

```diff
@@ -1,11 +1,63 @@
+from copy import deepcopy
+from typing import List
 from attrs import define, field
 
 
 @define
 class Checkpoint:
     """Checkpoint that counts the tokens in messages between the start and end index"""
 
-    start_index: int = field(default=-1)
-    end_index: int = field(default=0)
+    start_index: int = field(default=0)  # inclusive
+    end_index: int = field(default=0)  # inclusive
     token_count: int = field(default=0)
-    latest_generated_tokens: int = field(default=0)
+
+    def __deepcopy__(self, _) -> "Checkpoint":
+        """
+        Returns a deep copy of the Checkpoint object.
+        """
+        return Checkpoint(start_index=self.start_index, end_index=self.end_index, token_count=self.token_count)
+
+
+@define
+class CheckpointData:
+    """Aggregates all information about checkpoints"""
+
+    # the total number of tokens in the exchange. this is updated every time a checkpoint is
+    # added or removed
+    total_token_count: int = field(default=0)
+
+    # in order list of individual checkpoints in the exchange
+    checkpoints: List[Checkpoint] = field(factory=list)
+
+    # the offset to apply to the message index when calculating the last message index
+    # this is useful because messages on the exchange behave like a queue, where you can only
+    # pop from the left or right sides. This offset allows us to map the checkpoint indices
+    # to the correct message index, even if we have popped messages from the left side of
+    # the exchange in the past. we reset this offset to 0 when we empty the checkpoint data.
+    message_index_offset: int = field(default=0)
```

Comment:
I'm curious why this isn't on the exchange instead. If it were, then I feel this object is more tidy, an aggregation of checkpoints... Also, it would be a lot easier to understand its relationship to messages (as they are defined on the exchange). Food for thought and I'm sure I'm missing something.

Reply:
I totally agree! The `Exchange` class should be a clean representation of messages and checkpoints that stay in sync within it's same instance. However, our current implementation of the `Exchange` class is a frozen one - this means that variables such as  `message_index_offset` and `total_token_count` cannot be updated without hacky workarounds. Will be having a discussion with the team to hopefully change this.